src/gpu/align_z_to_other.cu

```diff
@@ -0,0 +1,239 @@
+        if (!(cudaFlags & cudaDeviceScheduleBlockingSync)) {
+            DEBUG("attempting to set the cudaDeviceScheduleBlockingSync flag");
+            std::lock_guard<std::mutex> lock(gpu_mutex);
+            // check flag again to make sure we don't reset it after some other thread set it for us
+            cudaCallErrorStatus = cudaGetDeviceFlags(&cudaFlags);
+            if (!(cudaFlags & cudaDeviceScheduleBlockingSync)) {
+                DEBUG("flag still not set");
+                cudaCallErrorStatus = cudaSetDeviceFlags(cudaDeviceScheduleBlockingSync);
+                if (cudaCallErrorStatus != cudaSuccess) {
+                    DEBUG("reseting device");
+                    cudaDeviceSynchronize();
+                    cudaDeviceReset();
+                    cudaCallErrorStatus = cudaSetDeviceFlags(cudaDeviceScheduleBlockingSync);
+                }
+            }
+        }
+
+
+        auto out_z = (uint16_t *)(z_aligned_to_other);
+
+        auto cudaDeleterUnsignedInt = [](unsigned int *ptr) {cudaFree(ptr);};
+        auto cudaDeleterUInt16t = [](uint16_t *ptr) {cudaFree(ptr);};
+        CudaStreamWrapper myCudaStream;
+        // Kernel invocation
+        dim3 threadsPerBlock(32, 32);
+        dim3 numBlocks(static_cast<int> (std::ceil(static_cast<float>(z_intrin.width) / threadsPerBlock.x)),
+                       static_cast<int>(std::ceil(
+                static_cast<float>(z_intrin.height) / threadsPerBlock.y)));
+        uint16_t * z_pixels_gpu;
+        unsigned int * temporary_out_pixels_gpu;
+        uint16_t * out_pixels_gpu_uint16t;
+
+
+        cudaCallErrorStatus = cudaMalloc((void **)&z_pixels_gpu, sizeof(uint16_t) * z_intrin.width * z_intrin.height);
+        std::unique_ptr<uint16_t, decltype(cudaDeleterUInt16t)> z_pixels_gpu_u_ptr(z_pixels_gpu, cudaDeleterUInt16t);
+        if (cudaSuccess != cudaCallErrorStatus || deviceCount == 0) {
+            std::cerr << "GPU processing failed while allocating space for z_pixels_gpu " << cudaGetErrorName(cudaCallErrorStatus) << " " << cudaGetErrorString(cudaCallErrorStatus) << std::endl;
+            return false;
+        }
+        cudaCallErrorStatus = cudaMalloc((void **)&temporary_out_pixels_gpu, sizeof(unsigned int) * other_intrin.width * other_intrin.height);
+        std::unique_ptr<unsigned int, decltype(cudaDeleterUnsignedInt)> out_pixels_gpu_u_ptr(temporary_out_pixels_gpu, cudaDeleterUnsignedInt);
+        if (cudaSuccess != cudaCallErrorStatus || deviceCount == 0) {
+            std::cerr << "GPU processing failed while allocating space for temporary_out_pixels_gpu " << cudaGetErrorName(cudaCallErrorStatus) << " " << cudaGetErrorString(cudaCallErrorStatus) << std::endl;
+            return false;
+        }
+        cudaCallErrorStatus = cudaMemsetAsync(temporary_out_pixels_gpu, sizeof(unsigned int) * other_intrin.width * other_intrin.height, 0, myCudaStream);
+        if (cudaSuccess != cudaCallErrorStatus) {
```

Comment:
Maybe wrap the calls in some helper to check return values?  Could clean up things a bit.

Reply:
yes i could implemnt (or steal :-) ) a cuda safe implementation. but we would still want to check for success....